runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingEngineComputationConfigFetcherTest.java

```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Test
+  public void testStart_requiresInitialConfig() throws IOException, InterruptedException {
+    WorkItem initialConfig =
+        new WorkItem()
+            .setJobId("job")
+            .setStreamingConfigTask(new StreamingConfigTask().setMaxWorkItemCommitBytes(10L));
+    CountDownLatch waitForInitialConfig = new CountDownLatch(1);
+    Set<StreamingEnginePipelineConfig> receivedPipelineConfig = new HashSet<>();
+    when(mockDataflowServiceClient.getGlobalStreamingConfigWorkItem())
+        .thenReturn(Optional.of(initialConfig));
+    streamingEngineConfigFetcher =
+        createConfigLoader(
+            /* waitForInitialConfig= */ true,
+            0,
+            config -> {
+              try {
+                receivedPipelineConfig.add(config);
+                waitForInitialConfig.await();
+              } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+              }
+            });
+    Thread asyncStartConfigLoader = new Thread(streamingEngineConfigFetcher::start);
+    asyncStartConfigLoader.start();
+    waitForInitialConfig.countDown();
+    asyncStartConfigLoader.join();
+    assertThat(receivedPipelineConfig)
+        .containsExactly(
+            StreamingEnginePipelineConfig.builder()
+                .setMaxWorkItemCommitBytes(
+                    initialConfig.getStreamingConfigTask().getMaxWorkItemCommitBytes())
+                .build());
+  }
+
+  @Test
+  public void testStart_startsPeriodicConfigRequests() throws IOException, InterruptedException {
+    WorkItem firstConfig =
+        new WorkItem()
+            .setJobId("job")
+            .setStreamingConfigTask(new StreamingConfigTask().setMaxWorkItemCommitBytes(10L));
+    WorkItem secondConfig =
+        new WorkItem()
+            .setJobId("job")
+            .setStreamingConfigTask(new StreamingConfigTask().setMaxWorkItemCommitBytes(15L));
+    WorkItem thirdConfig =
+        new WorkItem()
+            .setJobId("job")
+            .setStreamingConfigTask(new StreamingConfigTask().setMaxWorkItemCommitBytes(100L));
+    CountDownLatch numExpectedRefreshes = new CountDownLatch(3);
+    Set<StreamingEnginePipelineConfig> receivedPipelineConfig = new HashSet<>();
+    when(mockDataflowServiceClient.getGlobalStreamingConfigWorkItem())
+        .thenReturn(Optional.of(firstConfig))
+        .thenReturn(Optional.of(secondConfig))
+        .thenReturn(Optional.of(thirdConfig));
+
+    streamingEngineConfigFetcher =
+        createConfigLoader(
+            /* waitForInitialConfig= */ true,
+            Duration.millis(100).getMillis(),
+            config -> {
+              receivedPipelineConfig.add(config);
+              numExpectedRefreshes.countDown();
+            });
+
+    Thread asyncStartConfigLoader = new Thread(streamingEngineConfigFetcher::start);
+    asyncStartConfigLoader.start();
+    numExpectedRefreshes.await();
+    asyncStartConfigLoader.join();
+    assertThat(receivedPipelineConfig)
```

Comment:
seems safer to stop the fetcher's background thread before accessing received config (or use a concurrent set)

Reply:
done